feat: new documentation (#164)
tier4/CalibrationTools: sensor/docs/tutorials/tag_based_pnp_calibrator.md

# tag_based_pnp_calibrator

> In this tutorial, we will present a hands-on tutorial of the `tag_based_pnp_calibrator`. Although we provide pre-recorded rosbag, the flow of the tutorial is meant to show the user the steps they must perform in their own use cases with live sensors.

Author: The tense should be infinitive  here

Reviewer: @knzo25 
Sorry, I am not sure where to modify it. 
btw, it is the same context as the sfm one.

Author: @vividf 
Nice, fixed the sfm one in
d1a3bd2

General documentation regarding this calibrator can be found [here](../../tag_based_pnp_calibrator/README.md).

## Setup

This tutorial assumes that the user has already built the calibration tools.
Installation instructions can be found [here](../../README.md)

## Data preparation

Please download the data (rosbag) from [here](https://drive.google.com/drive/folders/1gFrrchW9mWM1huWMYuJ0nWB2n1BfCJag).

> The rosabg includes four different topics including `camera_info`, `image_rect_color/compressed`, `pointcloud_raw`, and `tf_static`.

Author: The provided rosbag

You can not say "including four topics including" and then list the four topics

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

Author: In your new version I think you are missing a ":"

Reviewer: sovled in https://github.com/tier4/CalibrationTools/pull/164/commits/62eabe1bc9a93c12cc6948924c8adeb58a36646d

## Environment preparation

### Overall calibration environment

> The required space for calibration depends on the vehicle and sensors used. For a normal consumer-level car, a space of `5m x 10m` should be sufficient.

Author: The space required depends on the lidar and cameras (fov and partially range)

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

### AprilTag (LidarTag)

> AprilTag are the only moving elements during the calibration process and must detected by both camera and lidar.

Author: The part of the only moving elements was required in the sfm calibrator.
It makes no sense here. Please do not copy text unless it is necessary

Reviewer: Got it! Thanks!
Fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

Depending on the lidar model and the available space, the required AprilTag size may vary, but so far we have had good results with 0.6m and 0.8m tags (the provided sizes correspond to an edge's size. In these cases the payloads are 0.45m and 0.6m).

In addition to the size of the AprilTag, which affects the physical positions where a tag can be detected, the structure on which the AprilTag is mounted is equally important. The shape and size of this structure may affect the lidar detection algorithm. Therefore, it is advisable to use a mount that positions the tag in such a way that it remains invisible to the sensor (refer to the provided example).

## Launching the tool

> In this tutorial, we will use the X2 vehicle of Tier IV.

Author: In this context, the X2 is not a vehicle. Actually, it never was

Reviewer: I only mentioned the x2 project this time.
Fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

First, run the sensor calibration manager:

```text
ros2 run sensor_calibration_manager sensor_calibration_manager
```

In `project`, select `x2`, and in `calibrator`, select `tag_based_pnp_calibrator`. Then, press `Continue`.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu1.jpg"  alt="menu1">
</p>

A menu titled `Launcher configuration` should appear in the UI, and the user may change any parameter he deems convenient.
For this tutorial, we will modify the default values `calibration_pairs` from `9` to `8` as the bag has 8 AprilTag detections and also modify the `camera_name` from `camera0` to `camera6`. After configuring the parameters, click `Launch`.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu2.jpg"  alt="menu2">
</p>

The following UI should be displayed. When the `Calibrate` button becomes available, click it.
If it does not become available, it means that either the required `tf` or services are not available.

> In this tutorial, since the `tf` are published by the provided rosbag, run the rag (`ros2 bag play camera_lidar.db3 --clock -r 0.1`) first and launch the tools afterward to trigger the `Calibrate` button.

Author: Trigger is not the correct word here

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

Author: In `In this tutorial, since the `tf` are published by the provided rosbag, start`
The use of the comma is incorrect

Reviewer: solved in https://github.com/tier4/CalibrationTools/pull/164/commits/14c627267650c7ca3b1a363a3ed4f2f5efa2fd48

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu3.jpg"  alt="menu3">
</p>

## Calibration

The calibration starts automatically after clicking the `Calibrate` button. It will keep calibrating the LidarTag detections and AprilTag detections until the number of the detections fits the user-defined `calibration_pairs` in the `Launcher configuration`.

When the user starts the calibration, `rviz` and the `image view` should be displayed like below.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization1.jpg"  alt="visualization1">
</p>

> After the tools detect the LidarTag and AprilTag, it will show the detection markers on the `rviz` and the `image view`. The text in the rviz will also display the current number of pairs of lidar detections and AprilTag detections.

Author: There is no consistency between lidar detections and ApriTag detections

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

Author: The text in the rviz
->
The text in rviz

The number of pairs of April....

Reviewer: fix in https://github.com/tier4/CalibrationTools/pull/164/commits/21825ba4a548b8bdb2e62a053fc93826296f85a1

Author: The new text is wrong.

`The text in rviz will also display the number of pairs of apriltag detections`
Does not the text display the number of `converged` pairs of `apriltag/lidartag` detections?

Reviewer: thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/f8893f6f66688b858006b76c8e030485a07630f7

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization2.jpg"  alt="visualization2">
</p>

> Once the user gets the converged detection, the user can start moving the tag to another position. Please make sure the moving distance is larger than the `calibration_min_pair_distance` and also make sure the tag is in the FOV of the lidar and camera.

Author: "Once the user gets the converged detection"
Users don't "get converged detections"

Author: You already said the user in the first part, so the second is redundant. Instead "he" can be used

Author: Of  both lidar and camera

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

At the end of the calibration, we can get 8 detection pairs which are shown below.

![segment](../images/tag_based_pnp_calibrator/visualization3.jpg)

## Results

> After the calibration process is finished, the `sensor_calibration_manager` will display the results in the tf tree and allow the user to save the calibration data to a file.

Author: After the calibration process finishes

Author: tf tree?
Refer to the proper elements in the widget, since there are several (that are tf trees)

Author: Also explain why and how the last two differ

Reviewer: Thanks! added the explanation in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu4.jpg" alt="menu4" width="500">
</p>

> The user can modify the `visualization options` on the right side of the `image view`. To compare the results, please set the `Marker size (m)` to `0.04` and set the `PC subsample factor` to `1`.

Author: Do they need to do this only at the end? 
Explain why should they need or could want to do this.
Why are we subsampling.
Compare results? Visualize results? Evaluate results?

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization_bar.jpg"  alt="visualization_bar" width="200">
</p>

> After setting the options above, change the `/initial_tf` (in the `visualization options`) to `/current_tf`. By doing this, it is easier to measure the difference after the calibration.

Author: It is easier it is not an explanation of the options.

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

> The images below show that with the calibrated transformation, the projected pointcloud aligns better with the image.

Author: Only better? We have quantitative results. Let's use them

Reviewer: Thanks! added console result in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

<table>
  <tr>
    <td><img src="../images/tag_based_pnp_calibrator/init_tf.jpg" alt="init_tf" width = 500px ></td>
    <td><img src="../images/tag_based_pnp_calibrator/calibrated_tf.jpg" alt="calibrated_tf" width = 500px ></td>
   </tr>
   <tr>
    <td><p style="text-align: center;">Before Calibration.</p></td>
    <td><p style="text-align: center;">After Calibration.</p></td>
  </tr>
</table>

## FAQ

> - Why the calibrator doesn't add calibration pairs?

Author: Why doesn't the calibrator add pairs

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

  1. One possible reason is that the current pair is too close to previously collected data. In that case, the current data is not accepted.

>   2. The timestamps of the lidar and camera are not synchronized, this can be checked with `ros2 topic echo [topic_name]`. Setting the parameter `use_receive_time` to `True` might help to solve the issue.

Author: Solves the issue? Always? It may work but are there any pitfalls?

Author: ros2 topic echo will print also the sensor data. --field header.stamp is better

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

>   3. The detections are not stable enough (detections don’t converge).

Author: If they are not stable enough what can they do?
What are the possible causes? Example: distance and resolution, wind, etc

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

Author: I think the comment was not addressed?
At least the commit does not indicate there was a change in the part I mention

Reviewer: I added the distance and wind at the 4. and 5.

Author: 4 and 5 are the causes of 3. There is no text that links this ideas
The tag being far from the camera is not the relevant factor (in most cases). It is the lidar

Reviewer: updated in commit https://github.com/tier4/CalibrationTools/pull/164/commits/bc9198c59221295e242cff68c01bbf1f79c2ec11.

I got this information from Sensor calibration説明会, if it is not enough for you, could you write them directly in the text? Thanks!!!

Author: Supplemented the explanation in [7fce3fb](https://github.com/tier4/CalibrationTools/pull/164/commits/7fce3fbe71709cf7410dd4c3034c03e7b8943e76)

- Why the UI doesn't launch?

  1. Check with `ros2 node list` if the relevant nodes have been started. It is possible that the provided parameters don’t match any of the valid arguments

>   2. If the UI crashes (see the console), it is probably due to bad PySide installation, invalid intrinsic parameters, or invalid extrinsic parameters.

Author: to a bad PySide2 installation
incorrect installation
etc

Author: In documentation, "see" usually means you are providing an example.
In this case, check is better, and even better would making obvious that we are referring to the user's local environment and not the tutorial

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

  3. The timestamps of the lidar and camera are not synchronized.

> - Why the reprojection errors are so high

Author: Other parts of this sections are presented as questions. Please make it consistent.

If detections look Ok and we have several of them, sure, intrinsics.
If we have outliers it also matches the high reprojection error

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

  1. Check whether the intrinsic parameters are correct.

- Why the reprojection error increases when more data is collected?

  1. When there are few samples, the model will fit the available data the best it can, even in the presence of noise (over-fitting). The more data is collected, the error may increase to a certain extent, but that is the model trying to fit all the data, being unable to fit the noise. However, it should reach a more-or-less table peak with about 10-15 pairs (depending on the data collection pattern/sampling).

- Why reprojection error does not seem low enough?

  1. The intrinsics may not be accurate, thus limiting the performance of the method.
  2. The boards are not appropriate (are bent).
  3. The boards moved too much while calibrating.

>   4. The lidar detections were not very good. Try collecting detections in areas where there is more resolution.

Author: Due to the structure you chose I do not know how to fit this in, but to avoid confusing the filtering algorithm, we require that the users moves decisively from one place to the other, not stopping and starting mid way.

Depending on the mount and the ground surface, the tags can oscillate quite a bit, which can be detrimental to the calibration process. In this case, the user can stop the tag with its hand. This should also be mentioned somewhere in this document

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

# tag_based_pnp_calibrator

> In this tutorial, we present a hands-on tutorial of the `tag_based_pnp_calibrator`. Although we provide pre-recorded rosbag, the flow of the tutorial is meant to show the user the steps they must perform in their own use cases with live sensors.

Author: `a pre-recorded rosbag`
Only one right?

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/cc97c92e86a0232697252ed721052b50e73c88a1

General documentation regarding this calibrator can be found [here](../../tag_based_pnp_calibrator/README.md).

## Setup

This tutorial assumes that the user has already built the calibration tools.
Installation instructions can be found [here](../../README.md)

## Data preparation

Please download the data (rosbag) from [here](https://drive.google.com/drive/folders/1gFrrchW9mWM1huWMYuJ0nWB2n1BfCJag).

The provided rosbag includes four different topics `camera_info`, `image_rect_color/compressed`, `pointcloud_raw`, and `tf_static`.

## Environment preparation

### Overall calibration environment

> The required space for calibration depends on the vehicle and sensors used. During the calibration, please always make sure that tags are detected by both camera and lidar. The user could always check whether the tags appear in the `rviz` and the `image view` UI.

Author: > The user could always check whether the tags appear in the `rviz` and the `image view` UI.

This is not well written.
For example: The user can check the detection status of the tags by observing rviz and the `image view` UI. 

Here there should also be an explanation, because mostly rviz and the UI show the converged pairs.
Rviz will show the current lidartag detections with cyan frames but how about the image detections?

Reviewer: Thanks!
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/cc75c68d22709c28e96567be1b5d47320469fb34

> The required space for calibration depends on the vehicle and sensors used. During the calibration, please always make sure that tags are detected by both camera and lidar. The user could always check whether the tags appear in the `rviz` and the `image view` UI. For instance, he can observe the cyan rhombus around the tag in `rviz` if the tag is successfully detected by lidar, as shown in the image below. For the camera, if the tag is visible in the `image view` UI, it should be detectable. So far, in our use cases, the tag is always detected as long as it is completely visible in the image.

Author: During calibration

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/5ae616b7380305967a763844e26df36c43895834

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/lidar_detection.jpg"  alt="lidar_detection">
</p>

> ### AprilTag (LidarTag)

Author: Ok, this time for sure it is not uniform :smile:

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/051490d47941ed5015842055cd4a8d926eab75c1

Author: That was not the problem...

Reviewer: Sorry, and Thanks!
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/77d046eaa7f35a85ab398ecda48a6f032b2e89cb

> In addition to the size of the AprilTag, which affects the physical positions where a tag can be detected, the structure on which the AprilTag is mounted is equally important. The shape and size of this structure may affect the lidar detection algorithm. Therefore, it is advisable to use a mount that positions the tag in such a way that it remains invisible to the sensor (refer to the provided example).

Author: The provided example in the main documentation, so it is not clear to what you refer here.
The images in this page do not show well the mount, but if you want to use them, please refer to those with some indexing

Reviewer: Thanks!
I removed them in https://github.com/tier4/CalibrationTools/pull/164/commits/cc75c68d22709c28e96567be1b5d47320469fb34.

Author: ` (so far, there have been no issues).`
This is not a useful explanation.
It could be: "so far, in our use cases, the tag is always detected as long as it is completely visible in the image"

Additionally, `Information can be found`
is not quite narutal:
`Details can be found` is better in my opinion

Author: @vividf

Reviewer: Thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/fd1c496905d40c494cfbf55c09b962dd863aac0e

## Launching the tool

> In this tutorial, we will use the X2 project as an example.

Author: We are not using future tense here

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/35f161a4b166734cf5b590d58c226ece36828b0d.

Note: the same issue also happened in the sfm docs.

Author: Thanks, mine was fixed in 2399a753b37e704565582a55066a0b936bbd7491

First, run the sensor calibration manager:

```text
ros2 run sensor_calibration_manager sensor_calibration_manager
```

In `project`, select `x2`, and in `calibrator`, select `tag_based_pnp_calibrator`. Then, press `Continue`.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu1.jpg"  alt="menu1">
</p>

A menu titled `Launcher configuration` should appear in the UI, and the user may change any parameter he deems convenient.

> For this tutorial, we will modify the default values `calibration_pairs` from `9` to `8` as the bag has 8 AprilTag detections and also modify the `camera_name` from `camera0` to `camera6`. After configuring the parameters, click `Launch`.

Author: We are not using future tense here

Author: bag -> rosbag

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/f8f0719c83cd1ae2f91600bace301adc4e34245f and https://github.com/tier4/CalibrationTools/pull/164/commits/3642f7523a8a6a1bdfca2efe07576b3398dcf540

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu2.jpg"  alt="menu2">
</p>

The following UI should be displayed. When the `Calibrate` button becomes available, click it.
If it does not become available, it means that either the required `tf` or services are not available.

> In this tutorial, since the `tf` are published by the provided rosbag, start by running the bag with the command `ros2 bag play camera_lidar.db3 --clock -r 0.1`. Afterward, launch the tools and click the `Calibrate` button."

Author: Extra "

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/14c627267650c7ca3b1a363a3ed4f2f5efa2fd48

    <img src="../images/tag_based_pnp_calibrator/menu3.jpg"  alt="menu3">
</p>

## Calibration

The calibration starts automatically after clicking the `Calibrate` button. It will keep calibrating the LidarTag detections and AprilTag detections until the number of the detections fits the user-defined `calibration_pairs` in the `Launcher configuration`.

When the user starts the calibration, `rviz` and the `image view` should be displayed like below.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization1.jpg"  alt="visualization1">
</p>

After the tools detect the LidarTag and AprilTag, it will show the detection markers on the `rviz` and the `image view`. The text in the rviz will also display the current number of pairs of LidarTag detections and AprilTag detections.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization2.jpg"  alt="visualization2">
</p>

Once the user observes the number of pairs increases in the `rviz`, he can start moving the tag to another position.

While moving the tag to another position, the user should move the tag decisively from one place to the other, not stopping and starting mid-way.

> Please make sure the moving distance is larger than the `calibration_min_pair_distance` and the tag is in the FOV of both lidar and camera.

Author: This one is not clear.
What you what to make the user understand is that the new position should be at least `calibration_min_pair_distance` from the closest detection in the set of converged detections

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/ea47fc19844e40b5e24e146866ed44b4d8bfae98

Author: We are getting close to the realm of nits, but there is a grammatical error here:

`Please make sure the new position should be at least`
to
`Please make sure the new position is at least`

Author: @vividf

Reviewer: thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/2338794d32e030eded8b97a2f9ee3d7e7917a1a8

Depending on the mount and the ground surface, the tags can oscillate quite a bit, which can be detrimental to the calibration process. In this case, the user can stop the tag with its hand.

At the end of the calibration, we can get 8 detection pairs which are shown below.

![tag_based_pnp_calibrator](../images/tag_based_pnp_calibrator/visualization3.jpg)

The output in the console should be as follows:

```text
[tag_based_pnp_calibrator_node]:  Initial reprojection error=15.41
[tag_based_pnp_calibrator_node]:  Current reprojection error=1.09
[tag_based_pnp_calibrator_node]:  Filtered reprojection error=1.10
```

## Results

After the calibration process finishes, the `sensor_calibration_manager` will display the results in the UI and allow the user to save the calibration data to a file.

> In the UI of the X2 project, three different TF trees are displayed: `Initial TF Tree`, `Calibration Tree`, and `Final TF Tree`. The `Initial TF Tree` presents the initial TF connections between sensors needed for calibration. The `Calibration Tree` shows the calibrated transformation between sensors, in this tutorial, `camera6/camera_optical_link` and `pandar_40p_front`. The `Final TF Tree` depicts the TF tree after incorporating the updated calibrated transformation. Since the transformations represented by the black arrows are fixed, the transformation between `front_unit_base_link` and `camera6/camera_link`, which is represented by the red arrow, can be calculated using the calibrated transformation.

Author: The documentation for `Final TF tree` is not quite correct

Author: "Since the transformations represented by the black arrows are fixed, the transformation between front_unit_base_link and camera6/camera_link, which is represented by the red arrow, can be calculated using the calibrated transformation."

is also kind of wrong. Please redact it again. For example it makes the user think black = fixed

Reviewer: @knzo25 
sry, could you explain them correctly, thanks.

I mentioned the black arrows (transformations) are fixed.

Author: > I mentioned the black arrows (transformations) are fixed.

Since in the figure are several black arrows that are not necessarily fixed you need to specify.
Also, you imply that because they are black they are fixes. That is not the case. The calibrator interface (.py file) determined as so. Why? That needs to be explained or it makes no sense. The color code actually means whether or not the connection changed. Please careful that you understand with what you write and that that is correct

Reviewer: Thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/e2210ae2139ca0627ac335674fa091113f8bee0c

Author: @vividf 
The explanation itself id good this time. However, the redaction is a little off so I will make a separate comment.

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/menu4.jpg" alt="menu4" width="500">
</p>

> During the calibration or at the end of the calibration, the user can modify the `visualization options` on the right side of the `image view`. Setting the `Marker size (m)` to `0.04` and setting the `PC subsample factor` to `1` can help the user visualize the projected pointcloud on the image.

Author: During calibration or during the calibration process.
Even then, using two times calibration in a raw sounds unnatural

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/0b72dce00835d172f6a268431c057d84afd04d8a

<p align="center">
    <img src="../images/tag_based_pnp_calibrator/visualization_bar.jpg"  alt="visualization_bar" width="200">
</p>

After setting the options above, change the `/initial_tf` (in the `visualization options`) to `/current_tf` to measure the difference after the calibration.

> The images below show that with the calibrated transformation, the projected pointcloud aligns with the image.

Author: ```suggestion
The images below show that with the calibrated transformation, the projected pointcloud aligns correctly with the image.
```

<table>
  <tr>
    <td><img src="../images/tag_based_pnp_calibrator/init_tf.jpg" alt="init_tf" width = 500px ></td>
    <td><img src="../images/tag_based_pnp_calibrator/calibrated_tf.jpg" alt="calibrated_tf" width = 500px ></td>
   </tr>
   <tr>
    <td><p style="text-align: center;">Before Calibration.</p></td>
    <td><p style="text-align: center;">After Calibration.</p></td>
  </tr>
</table>

## FAQ

- Why doesn't the calibrator add calibration pairs?

  1. One possible reason is that the current pair is too close to previously collected data. In that case, the current data is not accepted.

>   2. The timestamps of the lidar and camera are not synchronized, this can be checked with `ros2 topic echo [topic_name] --field header.stamp`. Setting the parameter `use_receive_time` to `True` might help to solve the issue.

Author: ```suggestion
  2. The timestamps of the lidar and camera are not synchronized, this can be checked with `ros2 topic echo [topic_name] --field header.stamp`. Setting the parameter `use_receive_time` to `True` might help to solve the issue but is not recommended as a long term solution.
```

  3. The detections are not stable enough (detections don’t converge).
  4. The position of the tag is too far away from the camera. The user can move the tag toward the camera until the detection shows on `rviz` or modify the parameter `max_tag_distance` in the launch file.
  5. The tag oscillates quite a bit due to the wind. The user can stop the tag with its hand.

- Why doesn't the UI launch?

  1. Check with `ros2 node list` if the relevant nodes have been started. It is possible that the provided parameters don’t match any of the valid arguments.
  2. If the UI crashes (check the console for details), it is probably due to a bad PySide installation, invalid intrinsic parameters, invalid extrinsic parameters, etc.
  3. The timestamps of the lidar and camera are not synchronized.

- Why does the reprojection error increase when more data is collected?

>   1. When there are few samples, the model will fit the available data the best it can, even in the presence of noise (over-fitting). The more data is collected, the error may increase to a certain extent, but that is the model trying to fit all the data, being unable to fit the noise. However, it should reach a more-or-less table peak with about 10-15 pairs (depending on the data collection pattern/sampling).

Author: ```suggestion
  1. When there are few samples, the model will fit the available data the best it can, even in the presence of noise (over-fitting). The more data is collected, the error may increase to a certain extent, but that corresponds to the model attempting to fit all the data, this time unable to fit the noise, resulting in a higher error. However, it should reach a more-or-less table peak with about 10-15 pairs (depending on the data collection pattern/sampling).
```
